airflow/contrib/operators/singularity_operator.py

```diff
@@ -0,0 +1,176 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from airflow.exceptions import AirflowException
+from airflow.models import BaseOperator
+from airflow.utils.decorators import apply_defaults
+from spython.main import Client
+import shutil
+import ast
+import os
+
+
+class SingularityOperator(BaseOperator):
+    """
+    Execute a command inside a Singularity container
+
+    Singularity has more seamless connection to the host than Docker, so
+    no special binds are needed to ensure binding content in the user $HOME
+    and temporary directories. If the user needs custom binds, this can
+    be done with --volumes
+
+    :param image: Singularity image or URI from which to create the container.
+    :type image: str
+    :param auto_remove: Delete the container when the process exits
+                        The default is False.
+    :type auto_remove: bool
+    :param command: Command to be run in the container. (templated)
+    :type command: str or list
+    :param start_command: start command to pass to the container instance
+    :type start_command: string or list
+    :param environment: Environment variables to set in the container. (templated)
+    :type environment: dict
+    :param force_pull: Pull the image on every run. Default is False.
+    :type force_pull: bool
+    :param volumes: List of volumes to mount into the container, e.g.
+        ``['/host/path:/container/path', '/host/path2:/container/path2']``.
+    :param options: other flags (list) to provide to the instance start
+    :type options: list
+    :param working_dir: Working directory to
+        set on the container (equivalent to the -w switch the docker client)
+    :type working_dir: str
+    """
+    template_fields = ('command', 'environment',)
+    template_ext = ('.sh', '.bash',)
+
+    @apply_defaults
+    def __init__(
+            self,
+            image,
+            api_version=None,
+            command=None,
+            start_command=None,
+            environment=None,
+            pull_folder=None,
+            force_pull=False,
+            volumes=None,
+            options=None,
+            working_dir=None,
+            auto_remove=False,
```

Comment:
Would you mind adding type annotations? It makes the code easier to understand and Airflow is using mypy for static analysis.

Reply:
I don't see examples for other contrib operators (just looked at Docker) and I don't have experience using them, so I need to ask for help here. I can see that the imports are generally `from typing import ...` but I'm not sure how to use them.

Reply:
Type hints allow us to easily indicate what types arguments have and if they are necessary:
https://github.com/apache/airflow/blob/e2524fac59df1a405a38f7f71b99436b44a47378/airflow/providers/google/cloud/operators/pubsub.py#L113-L128

In above example you can see that `labels` argument is `Optional` and if passed should be the type of `Dict[str, str]` meaning that all keys and values should be strings. It's really easy to add and helps a lot (no need to search for info in docstring). More information here: https://docs.python.org/3/library/typing.html

Reply:
okay I'll give this a shot.

Reply:
okay, added! I really appreciate the quick examples, they made it much easier to digest the new syntax fairly quickly.